File: sdks/typescript/src/crypto/subject.ts
/**
 * Subject Keys Cryptographic Utilities
 * Based on docs/SUBJECT-KEYS-ARCHITECTURE.md
 *
 * Provides client-side HMAC derivation for pseudonymous subject identification.
 * The server never sees raw subject identifiers (customer IDs, user IDs, etc.).
 */

import sodium from 'libsodium-wrappers';

/**
 * Subject ID type: 64-character hex string (HMAC-SHA512/256 output)
 */
export type SubjectId = string;

/**
 * Memory scope for subject isolation
 */
export type MemoryScope = 'SUBJECT' | 'SEGMENT' | 'GLOBAL';

/**
 * Subject context for memory operations
 */
export interface SubjectContext {
  /** HMAC-derived pseudonymous subject identifier (64 hex chars) */
  subjectId?: SubjectId;
  /** Memory visibility scope */
  scope: MemoryScope;
  /** Segment identifier for SEGMENT-scoped memories */
  segmentId?: string;
  /** Enterprise tenant identifier */
  tenantId?: string;
}

/**
 * Subject derivation options
 */
export interface SubjectDerivationOptions {
  /** Custom domain separator (default: 'xache:subject:v1') */
  domain?: string;
}

/**
 * Ensure libsodium is ready
 */
async function ensureSodiumReady(): Promise<void> {
  await sodium.ready;
}

/**
 * Derives a pseudonymous subject ID using HMAC-SHA512/256.
 *
 * This function creates a deterministic, irreversible identifier from a raw subject ID
 * (such as a customer ID or user ID). The derived subject ID:
 * - Is pseudonymous (cannot be reversed to the original ID)
 * - Is deterministic (same inputs always produce same output)
 * - Is unique per agent (different agents produce different subject IDs)
 * - Is 64 hex characters (256 bits of entropy)
 *
 * @param agentKey - Agent's encryption key (32 bytes, base64)
 * @param rawSubjectId - Raw subject identifier (customer ID, user ID, etc.)
 * @param options - Optional derivation configuration
 * @returns 64-character hex string (pseudonymous subject ID)
 *
 * @example
 * ```typescript
 * const subjectId = await deriveSubjectId(agentKey, 'customer_12345');
 * // Returns: 'a1b2c3d4e5f6...' (64 hex chars)
 * ```
 */
export async function deriveSubjectId(
  agentKey: string,
  rawSubjectId: string,
  options?: SubjectDerivationOptions
): Promise<SubjectId> {
  await ensureSodiumReady();

  // Decode the agent key from base64
  const keyBytes = sodium.from_base64(agentKey);

  if (keyBytes.length !== 32) {
    throw new Error(`Invalid agent key length: expected 32 bytes, got ${keyBytes.length}`);
  }

  // Create domain-separated input to prevent cross-protocol attacks
  const domain = options?.domain ?? 'xache:subject:v1';
  const input = `${domain}:${rawSubjectId}`;
  const inputBytes = new TextEncoder().encode(input);

  // Use HMAC-SHA512/256 (first 256 bits of HMAC-SHA512)
  // This provides 256-bit security while being efficient
  const hmacKey = sodium.crypto_generichash(32, keyBytes, undefined);
  const hmac = sodium.crypto_auth(inputBytes, hmacKey);

  // Convert to hex string (64 characters for 32 bytes)
  return sodium.to_hex(hmac);
}

/**
 * Validates that a subject ID has the correct format.
 *
 * @param subjectId - Subject ID to validate
 * @returns true if valid, false otherwise
 */
export function isValidSubjectId(subjectId: string): boolean {
  if (typeof subjectId !== 'string') {
    return false;
  }
  // Must be exactly 64 hex characters
  return /^[a-f0-9]{64}$/.test(subjectId);
}

/**
 * Validates that a scope is valid.
 *
 * @param scope - Scope to validate
 * @returns true if valid, false otherwise
 */
export function isValidScope(scope: string): scope is MemoryScope {
  return scope === 'SUBJECT' || scope === 'SEGMENT' || scope === 'GLOBAL';
}

/**
 * Creates a SUBJECT-scoped context.
 *
 * @param subjectId - HMAC-derived subject identifier
 * @param tenantId - Optional tenant identifier
 * @returns Subject context for SUBJECT scope
 */
export function createSubjectContext(
  subjectId: SubjectId,
  tenantId?: string
): SubjectContext {
  if (!isValidSubjectId(subjectId)) {
    throw new Error('Invalid subject ID format: must be 64 hex characters');
  }
  return {
    subjectId,
    scope: 'SUBJECT',
    tenantId,
  };
}

/**
 * Creates a SEGMENT-scoped context.
 *
 * @param segmentId - Segment identifier
 * @param tenantId - Optional tenant identifier
 * @returns Subject context for SEGMENT scope
 */
export function createSegmentContext(
  segmentId: string,
  tenantId?: string
): SubjectContext {
  if (!segmentId || segmentId.length === 0) {
    throw new Error('Segment ID cannot be empty');
  }
  return {
    segmentId,
    scope: 'SEGMENT',
    tenantId,
  };
}

/**
 * Creates a GLOBAL-scoped context.
 *
 * @param tenantId - Optional tenant identifier
 * @returns Subject context for GLOBAL scope
 */
export function createGlobalContext(tenantId?: string): SubjectContext {
  return {
    scope: 'GLOBAL',
    tenantId,
  };
}

/**
 * Validates a subject context for consistency.
 *
 * @param context - Subject context to validate
 * @throws Error if context is invalid
 */
export function validateSubjectContext(context: SubjectContext): void {
  if (!isValidScope(context.scope)) {
    throw new Error(`Invalid scope: ${context.scope}`);
  }

  if (context.scope === 'SUBJECT') {
    if (!context.subjectId) {
      throw new Error('subjectId is required when scope is SUBJECT');
    }
    if (!isValidSubjectId(context.subjectId)) {
      throw new Error('Invalid subjectId format: must be 64 hex characters');
    }
  }

  if (context.scope === 'SEGMENT') {
    if (!context.segmentId) {
      throw new Error('segmentId is required when scope is SEGMENT');
    }
    if (context.segmentId.length === 0) {
      throw new Error('segmentId cannot be empty');
    }
  }
}

/**
 * Batch derives subject IDs for multiple raw identifiers.
 * More efficient than calling deriveSubjectId multiple times.
 *
 * @param agentKey - Agent's encryption key (32 bytes, base64)
 * @param rawSubjectIds - Array of raw subject identifiers
 * @param options - Optional derivation configuration
 * @returns Map of raw ID to derived subject ID
 */
export async function batchDeriveSubjectIds(
  agentKey: string,
  rawSubjectIds: string[],
  options?: SubjectDerivationOptions
): Promise<Map<string, SubjectId>> {
  await ensureSodiumReady();

  const results = new Map<string, SubjectId>();

  // Decode key once
  const keyBytes = sodium.from_base64(agentKey);
  if (keyBytes.length !== 32) {
    throw new Error(`Invalid agent key length: expected 32 bytes, got ${keyBytes.length}`);
  }

  const domain = options?.domain ?? 'xache:subject:v1';
  const hmacKey = sodium.crypto_generichash(32, keyBytes, undefined);

  for (const rawId of rawSubjectIds) {
    const input = `${domain}:${rawId}`;
    const inputBytes = new TextEncoder().encode(input);
    const hmac = sodium.crypto_auth(inputBytes, hmacKey);
    results.set(rawId, sodium.to_hex(hmac));
  }

  return results;
}

File: sdks/python/xache/crypto/subject.py
"""
Subject Keys Cryptographic Utilities

Provides client-side HMAC derivation for pseudonymous subject identification.
The server never sees raw subject identifiers (customer IDs, user IDs, etc.).
"""

import hmac
import hashlib
import re
from typing import Dict, List, Literal, Optional

from dataclasses import dataclass


# Types
SubjectId = str  # 64-character hex string (HMAC-SHA256 output)
MemoryScope = Literal["SUBJECT", "SEGMENT", "GLOBAL"]
SegmentId = str
TenantId = str


@dataclass
class SubjectContext:
    """Subject context for memory operations."""
    scope: MemoryScope
    subject_id: Optional[SubjectId] = None
    segment_id: Optional[SegmentId] = None
    tenant_id: Optional[TenantId] = None


@dataclass
class SubjectRetrievalContext:
    """Subject context for memory retrieval operations."""
    subject_id: Optional[SubjectId] = None
    include_segment: bool = True
    include_global: bool = True
    segment_id: Optional[SegmentId] = None
    tenant_id: Optional[TenantId] = None


@dataclass
class SubjectDerivationOptions:
    """Subject derivation options."""
    domain: str = "xache:subject:v1"


def derive_subject_id(
    agent_key: bytes,
    raw_subject_id: str,
    options: Optional[SubjectDerivationOptions] = None,
) -> SubjectId:
    """
    Derive a pseudonymous subject ID using HMAC-SHA256.

    Creates a deterministic, irreversible identifier from a raw subject ID.
    The derived subject ID:
    - Is pseudonymous (cannot be reversed to the original ID)
    - Is deterministic (same inputs always produce same output)
    - Is unique per agent (different agents produce different subject IDs)
    - Is 64 hex characters (256 bits)

    Args:
        agent_key: Agent's encryption key (32 bytes)
        raw_subject_id: Raw subject identifier (customer ID, user ID, etc.)
        options: Optional derivation configuration

    Returns:
        64-character hex string (pseudonymous subject ID)
    """
    if len(agent_key) != 32:
        raise ValueError(f"Invalid agent key length: expected 32 bytes, got {len(agent_key)}")

    domain = options.domain if options else "xache:subject:v1"
    message = f"{domain}:{raw_subject_id}".encode("utf-8")

    # Derive HMAC key from agent key using BLAKE2b
    hmac_key = hashlib.blake2b(agent_key, digest_size=32).digest()
    mac = hmac.new(hmac_key, message, hashlib.sha256)
    return mac.hexdigest()


def is_valid_subject_id(subject_id: str) -> bool:
    """Validate that a subject ID has the correct format (64 hex chars)."""
    if not isinstance(subject_id, str):
        return False
    return bool(re.match(r"^[a-f0-9]{64}$", subject_id))


def is_valid_scope(scope: str) -> bool:
    """Validate that a scope is valid."""
    return scope in ("SUBJECT", "SEGMENT", "GLOBAL")


def create_subject_context(
    subject_id: SubjectId,
    tenant_id: Optional[str] = None,
) -> SubjectContext:
    """Create a SUBJECT-scoped context."""
    if not is_valid_subject_id(subject_id):
        raise ValueError("Invalid subject ID format: must be 64 hex characters")
    return SubjectContext(
        subject_id=subject_id,
        scope="SUBJECT",
        tenant_id=tenant_id,
    )


def create_segment_context(
    segment_id: str,
    tenant_id: Optional[str] = None,
) -> SubjectContext:
    """Create a SEGMENT-scoped context."""
    if not segment_id:
        raise ValueError("Segment ID cannot be empty")
    return SubjectContext(
        segment_id=segment_id,
        scope="SEGMENT",
        tenant_id=tenant_id,
    )


def create_global_context(
    tenant_id: Optional[str] = None,
) -> SubjectContext:
    """Create a GLOBAL-scoped context."""
    return SubjectContext(
        scope="GLOBAL",
        tenant_id=tenant_id,
    )


def validate_subject_context(context: SubjectContext) -> None:
    """
    Validate a subject context for consistency.

    Raises:
        ValueError: If context is invalid
    """
    if not is_valid_scope(context.scope):
        raise ValueError(f"Invalid scope: {context.scope}")

    if context.scope == "SUBJECT":
        if not context.subject_id:
            raise ValueError("subject_id is required when scope is SUBJECT")
        if not is_valid_subject_id(context.subject_id):
            raise ValueError("Invalid subject_id format: must be 64 hex characters")

    if context.scope == "SEGMENT":
        if not context.segment_id:
            raise ValueError("segment_id is required when scope is SEGMENT")


def derive_entity_key(agent_key: bytes, entity_name: str) -> SubjectId:
    """
    Derive an HMAC entity key for knowledge graph entities.

    Entity names are normalized (lowercase, trimmed) before HMAC so
    "Alice Chen" and "alice chen" produce the same key.

    Args:
        agent_key: Agent's encryption key (32 bytes)
        entity_name: Raw entity name

    Returns:
        64-character hex string
    """
    normalized = entity_name.strip().lower()
    return derive_subject_id(
        agent_key, normalized,
        SubjectDerivationOptions(domain="xache:entity:v1"),
    )


def batch_derive_subject_ids(
    agent_key: bytes,
    raw_subject_ids: List[str],
    options: Optional[SubjectDerivationOptions] = None,
) -> Dict[str, SubjectId]:
    """
    Batch derive subject IDs for multiple raw identifiers.

    Args:
        agent_key: Agent's encryption key (32 bytes)
        raw_subject_ids: Array of raw subject identifiers
        options: Optional derivation configuration

    Returns:
        Dict mapping raw ID to derived subject ID
    """
    if len(agent_key) != 32:
        raise ValueError(f"Invalid agent key length: expected 32 bytes, got {len(agent_key)}")

    domain = options.domain if options else "xache:subject:v1"
    hmac_key = hashlib.blake2b(agent_key, digest_size=32).digest()

    results: Dict[str, SubjectId] = {}
    for raw_id in raw_subject_ids:
        message = f"{domain}:{raw_id}".encode("utf-8")
        mac = hmac.new(hmac_key, message, hashlib.sha256)
        results[raw_id] = mac.hexdigest()

    return results


def batch_derive_entity_keys(
    agent_key: bytes,
    entity_names: List[str],
) -> Dict[str, SubjectId]:
    """
    Batch derive entity keys for multiple entity names.

    Args:
        agent_key: Agent's encryption key (32 bytes)
        entity_names: Array of entity names

    Returns:
        Dict mapping normalized name to derived entity key
    """
    if len(agent_key) != 32:
        raise ValueError(f"Invalid agent key length: expected 32 bytes, got {len(agent_key)}")

    domain = "xache:entity:v1"
    hmac_key = hashlib.blake2b(agent_key, digest_size=32).digest()

    results: Dict[str, SubjectId] = {}
    for name in entity_names:
        normalized = name.strip().lower()
        message = f"{domain}:{normalized}".encode("utf-8")
        mac = hmac.new(hmac_key, message, hashlib.sha256)
        results[normalized] = mac.hexdigest()

    return results
